src/OrleansGetEventStore/GetEventStoreProvider.cs

```diff
@@ -0,0 +1,234 @@
+            var hostAddress = Dns.GetHostAddresses(hostName).First(a => a.AddressFamily == AddressFamily.InterNetwork);
+            
+            Log.Info($"Init Name={Name} ConnectionString={connectionString} Severity={Log.SeverityLevel}");
+
+            var username = config.GetProperty(UsernameParameterName, "admin");
+            var password = config.GetProperty(PasswordParameterName, "changeit");
+
+            var settings = ConnectionSettings.Create()
+                .KeepReconnecting().KeepRetrying()
+                .SetDefaultUserCredentials(new UserCredentials(username, password));
+
+            this.Connection = EventStoreConnection.Create(settings, new IPEndPoint(hostAddress, hostPort));
+            await this.Connection.ConnectAsync();
+
+            Log.Verbose("Successfully Connected");
+        }
+
+
+        public override Task Close()
+        {
+            this.Connection.Close();
+            return TaskDone.Done;
+        }
+
+        public override string DefaultStreamName(Type grainType, GrainReference grainReference)
+        {
+            return string.Format("{0}_{1}", serviceId, grainReference.ToKeyString());
+        }
+
+        private static Dictionary<string,object> emptyCommitHeaders = new Dictionary<string, object>();
+
+        public override async Task<bool> AppendToStream(string streamName, IEnumerable<KeyValuePair<Guid,object>> events, int? expectedVersion)
+        {
+            var writePageSize = 500;
+
+            var eventsToSave = events.Select(e => ToEventData(e.Key, e.Value, emptyCommitHeaders)).ToArray();
+
+            if (eventsToSave.Length == 0)
+                return true;
+
+            var version = expectedVersion.HasValue
+                ? expectedVersion.Value == 0
+                    ? ExpectedVersion.NoStream
+                    : expectedVersion.Value - 1
+                : ExpectedVersion.Any;
+
+            if (eventsToSave.Length < writePageSize)
+            {
+                try
+                {
+                    await this.Connection.AppendToStreamAsync(streamName, version, eventsToSave);
+                }
+                catch (WrongExpectedVersionException)
+                {
+                    return false;
+                }
+            }
+            else
+            {
+                try
+                {
+                    var transaction = await this.Connection.StartTransactionAsync(streamName, version);
+
+                    var position = 0;
+                    while (position < eventsToSave.Length)
+                    {
+                        var pageEvents = eventsToSave.Skip(position).Take(writePageSize);
+                        await transaction.WriteAsync(pageEvents);
+                        position += writePageSize;
+                    }
+
+                    await transaction.CommitAsync();
+                }
+                catch (WrongExpectedVersionException)
```

Comment:
It's handled automatically. Exactly what happens depends on whether this was a conditional or an unconditional event (i.e. whether the grain called `RaiseEvent` or `RaiseConditionalEvent`).

Unconditional events are automatically retried until successful. Conditional events fail (the task returned by `RaiseConditionalEvent` resolves to false), because it means they lost a race against some other event.

This is documented on the "Replicated Instances" page.